quickwit/quickwit-ingest/src/ingest_v2/fetch.rs

```diff
@@ -364,58 +373,94 @@ fn select_preferred_and_failover_ingesters(
     }
 }
 
-/// Streams records from the preferred ingester and fails over to the other ingester if an error
-/// occurs.
+/// Performs multiple fault-tolerant fetch stream attempts until the stream reaches
+/// the end of the shard.
 #[allow(clippy::too_many_arguments)]
-async fn fault_tolerant_fetch_stream_task(
+async fn retrying_fetch_stream(
     client_id: String,
     index_uid: IndexUid,
     source_id: SourceId,
     shard_id: ShardId,
     mut from_position_exclusive: Position,
     ingester_ids: Vec<NodeId>,
     ingester_pool: IngesterPool,
+    retry_params: RetryParams,
+    fetch_response_tx: mpsc::Sender<Result<FetchResponseV2, FetchStreamError>>,
+) {
+    let mut num_attempts = 0;
+
+    loop {
+        fault_tolerant_fetch_stream(
+            client_id.clone(),
+            index_uid.clone(),
+            source_id.clone(),
+            shard_id,
+            &mut from_position_exclusive,
+            &ingester_ids,
+            ingester_pool.clone(),
+            fetch_response_tx.clone(),
+        )
+        .await;
+
+        num_attempts += 1;
+
+        if num_attempts >= retry_params.max_attempts || from_position_exclusive == Position::Eof {
+            break;
+        }
+        let delay = retry_params.compute_delay(num_attempts);
+        tokio::time::sleep(delay).await;
+    }
```

Comment:
```suggestion
    for attempt in 0..retry_params.max_attempts {
        fault_tolerant_fetch_stream(
            client_id.clone(),
            index_uid.clone(),
            source_id.clone(),
            shard_id,
            &mut from_position_exclusive,
            &ingester_ids,
            ingester_pool.clone(),
            fetch_response_tx.clone(),
        )
        .await;
        if from_position_exclusive == Position::Eof {
            break;
        }
        let delay = retry_params.compute_delay(attempt);
        tokio::time::sleep(delay).await;
    }
```

Reply:
It might seem stilly, but `for` loop make it  trivial to proofread that your program terminates.